[cc65] Fixed constant boolean expressions

acqn opened the pull request:
(8/8) commits.
- [x] Moved the almighty `bug250.c` from `test/misc` to `test/val`.
  - [x] Added integer boolean type string - no longer set the "expression tested" flag with constant results in comparison.
  - [x] Fixed #1123 - no more warnings on "const comparisons" that are desired in constant context.
  - [x] Warnings on expressions with no effects instead of statements with no effects.
  - [x] Fixed #250 (the OP part) - const ternary operator.
  - [x] Fixed #1221 - ternary couldn't use struct/union types as the true/false-branch expressions.
  - [x] Fixed #250 (https://github.com/cc65/cc65/issues/250#issuecomment-658003430) - logical AND and logical OR.
- [x] Added a constant AND/OR codegen test.

---
It might be worth mentioning that despite 1d7fcc67a498f65a4687099b87a40d402c039c78 saves nearly half of the labels, the new testcase d4025efaba798c652302bb49fb45506f8504b5c0 still easily reaches a big label number (>= L1131). The enhancements suggested in #301 might be desired.

acqn commented:
I think I really hate `test/misc`.

acqn commented:
Huh? So Travis CI doesn't respect the order in which the commits are pushed, but just reschedules the tests according to the "timestamps" of the commits? @oliverschmidt

acqn commented:
I give up. Enjoy the commit fusion.

acqn commented:
It seems to me that the "random" build failure tends to happen when there are multiple tests scheduled. I don't know if that indicates some hidden memory corruption bug since I can't reproduce it on my PC.

oliverschmidt commented:
> Huh? So Travis CI doesn't respect the order in which the commits are pushed, but just reschedules the tests according to the "timestamps" of the commits? @oliverschmidt

Either I get you wrong or I'd be pretty surprised if it were as you describe. The picture I have in mind is that Travis CI doesn't know anything about Git. When something is pushed to the upstream repo, a test of what was pushed is triggered. However, If your experience contradicts my understanding then my understanding is likely wrong...

jmr commented:
It says it builds the merge (so not exactly what was pushed if HEAD is different), but I'm not sure if that helps to explain anything.
https://docs.travis-ci.com/user/pull-requests/#how-pull-requests-are-built

oliverschmidt commented:
> It seems to me that the "random" CI build failure (https://travis-ci.org/github/cc65/cc65/jobs/721197860) tends to happen when there are multiple tests scheduled. I don't know if that indicates some hidden memory corruption bug or not since I can't reproduce it on my PC.

I re-ran that Travis CI job quite some times. It succeeded each time.

acqn commented:
> 
> 
> > It seems to me that the "random" CI build failure (https://travis-ci.org/github/cc65/cc65/jobs/721197860) tends to happen when there are multiple tests scheduled. I don't know if that indicates some hidden memory corruption bug or not since I can't reproduce it on my PC.
> 
> I re-ran that Travis CI job quite some times. It succeeded each time.

Could you re-run https://travis-ci.org/github/cc65/cc65/jobs/721079643?

acqn commented:
> 
> 
> It says it builds the merge (so not exactly what was pushed if HEAD is different), but I'm not sure if that helps to explain anything.
> https://docs.travis-ci.com/user/pull-requests/#how-pull-requests-are-built

It appears to me that the CI "caches" the results of some _commits_ (instead of "merged snapshots") and reuses them without redoing the tests.

acqn commented:
True order:
1. Added integer boolean type string. 17bbba7
2. Moved bug250.c from test/misc as it is to be fixed. b15ab34
3. Fixed warnings on const comparison. 3ea3887
4. Moved bug250.c from test/misc as it is to be fixed. b15ab34
5. Fixed constant expression checks with no-code requirement. 7255111
6. Fixed logical AND and logical OR. 85ea066
7. Avoided generating unnecessary true-case labels in logic AND/OR. f289ea6
8. Added test/ref/pr1220 for testing constant AND/OR code generation. b6a4715

oliverschmidt commented:
> Could you re-run https://travis-ci.org/github/cc65/cc65/jobs/721079643? That one was weird that it fore-triggered a bug pushed after the commits it was testing.

I re-ran that Travis CI job quite some times. It succeeded each time.

oliverschmidt commented:
> I give up. I don't want to fight test/misc, GitHub and Travis CI at the same time. 

If I understand things right, Travis CI entered the arena because the cc65 build wasn't fully parallelizable. Therefore I worked on that issue. I now closed https://github.com/cc65/cc65/issues/1080 and presume that you can therefore revert to a local build.

acqn commented:
Could it be possible that the failed tests were caused by flawed cc65 executable builds? The only way a revision test could demonstrate a "future" bug would be:
1. There was a different hidden bug that could demenstrate exactly the same symptoms.
2. The test was indeed done with the future commit built into the executable.

jmr commented on src/cc65/datatype.c:
```diff
@@ -66,6 +66,7 @@ Type type_int[]         = { TYPE(T_INT),    TYPE(T_END) };
 Type type_uint[]        = { TYPE(T_UINT),   TYPE(T_END) };
 Type type_long[]        = { TYPE(T_LONG),   TYPE(T_END) };
 Type type_ulong[]       = { TYPE(T_ULONG),  TYPE(T_END) };
+Type type_bool[]        = { TYPE(T_INT),    TYPE(T_END) };
```
What should happen to `type_bool` if `_Bool` is introduced? Then it becomes `TYPE(T_INT)` or we just need to figure it out later?

acqn replied:
It would use a new `T_BOOL` (that is  `T_TYPE_BOOL | T_CLASS_INT | T_SIGN_UNSIGNED | T_SIZE_BOOL`, and probably a `BOOL_BITS` that equals `1`. It would need special handling like `char` and `enum`.

jmr commented on src/cc65/exprdesc.h:
```diff
@@ -104,29 +104,58 @@ enum {
+    E_EVAL_CONST        = 0x040000,     /* Result must be immutable */
+    E_EVAL_COMPILE_TIME = 0x0C0000,     /* Result must be known at compile time */
+    E_EVAL_PURE         = 0x100000,     /* Evaluation must have no side effects */
+    E_EVAL_STATIC       = 0x340000,     /* Evaluation must generate no code */
```
I think having `STATIC` and `CONST` names that mean different things from `static` and `const` in C is a bit confusing.

acqn replied:
There have been different meanings of `const`/`constant` in cc65 for a long time, which I think should be enough to keep people aware of that. And `static` in C mostly refers to object linkage rather than expression traits, except for `_Static_assert`, which exactly indicates the same "no runtime" virtual as used here.

Anyways, do you have suggestions for better names? `E_EVAL_IMMUTABLE` could be  good for `E_EVAL_CONST`, but `E_EVAL_NO_CODE`, `E_EVAL_CODELESS` or `E_EVAL_EMPTY_CODE` for `E_EVAL_STATIC`?

jmr replied:
I didn't realize, but many/most of these are currently unused.

It would definitely help to expand the comments. "must be immutable" doesn't tell me much. Is `(i++, 0)` immutable? Having good documentation is more important than getting the right names right now. 

`E_EVAL_NO_CODE` seems good for `E_EVAL_STATIC`. I'll try to think about the other names.

acqn replied:
The comment says "**Result** must be immutable". The result of the expression `(i++, 0)` is always `0`. Thus it is immutable.

jmr commented on src/cc65/expr.h:
```diff
@@ -54,23 +54,23 @@ int evalexpr (unsigned flags, void (*Func) (ExprDesc*), ExprDesc* Expr);
-** a constant numeric integer value, a diagnostic will be printed, and the
+ExprDesc StaticConstExpr (void (*Func) (ExprDesc*));
+/* Get an expression evaluated via the given function. If the result is not a
+** static constant expression, a diagnostic will be printed, and the value is
```
What is a "static constant expression"? Explain or include pointers to which definitions of these terms you're using.

acqn replied:
It could be read at the definition of `E_EVAL_C_CONST`. `Static` in such a context means "evaluated without any runtime code generated" as in #1196. What about changing the phrase
> static constant expression

to
> constant expression needing no runtime code

?

oliverschmidt commented:
@acqn: Do you consider the two remarks from @jmr resolved? If yes, please mark them as such.

acqn commented:
Not yet. I can rename them, but I haven't come up with good names.

jmr commented on src/cc65/expr.h:
```diff
@@ -54,25 +54,25 @@ int evalexpr (unsigned flags, void (*Func) (ExprDesc*), ExprDesc* Expr);
-** errors that result from this input error.
+ExprDesc StaticConstExpr (void (*Func) (ExprDesc*));
+/* Get an expression evaluated via the given function. If the result is not a
+** constant expression without runtime code generated, a diagnostic will be
```
Can you say something like "constant expression as defined by `E_EVAL_...`?

acqn replied:
What about both?

jmr commented on src/cc65/exprdesc.h:
```diff
@@ -104,29 +104,58 @@ enum {
+    E_EVAL_NONE         = 0x000000,     /* No requirements */
+    E_EVAL_CONST        = 0x040000,     /* Result must be immutable */
+    E_EVAL_COMPILE_TIME = 0x0C0000,     /* Result must be known at compile time */
+    E_EVAL_PURE         = 0x100000,     /* Evaluation must have no side effects */
```
Pure usually means no side effects and same result for same inputs. `E_EVAL_NO_SIDE_EFFECTS` is a bit long, but `*p` isn't pure, is it?

acqn replied:
The concept "Pure" was borrowed from "pure function" but it is about expressions, not just functions. It's true that the names could have better substitutes. I'll go with `E_EVAL_NO_SIDE_EFFECTS`.

jmr commented:
Thanks! The new comments are helpful.

acqn commented:
There was a potential subexpression issue in ternary operation similar to #1181. Now fixed in 74def4608aab0065d118e4a072e8aab9b6918cd6.

acqn commented:
There were two bugs with 74def4608aab0065d118e4a072e8aab9b6918cd6. Incoming hotfix.
